### PR TITLE
Add advertiser split testing for Brave ads

### PR DESCRIPTION
**Pull request by moritzhaller**
Resolves https://github.com/brave/brave-browser/issues/14613

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Run with `npm start -- --enable-logging=stderr --vmodule="*/variations/*"=6,"*/bat-native-ads/*"=6,"*/brave_ads/*"=6,"*/brave_user_model/*"=6,"*/bat_ads/*"=6 --use-dev-goupdater-url --brave-ads-staging --rewards=staging=true --variations-server-url=https://variations.bravesoftware.com/seed --fake-variations-channel=beta`

- Inject test catalog that has a creative set with field `"splitTestGroup": "GroupA"` set.
- Start browser without enabling ads and quit. Then start a second time with enabling ads (**This is important since `seed` will only be initialised on restart!**)
- Verify in logs that study loaded correctly with `[10210:775:0313/084147.162484:VERBOSE1:study_filtering.cc(346)] Kept study AdvertiserSplitTestingStudy.`
- trigger ad and verify it is being served and not filtered out.
- Change the test catalog to contain the field `"splitTestGroup": "GroupB"` and wipe profile.
- Re-run test and verify the frequency cap is now excluding this ad.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ad_server/get_catalog_url_request_builder.cc [hunk 1]
```diff
@@ -27,6 +27,7 @@ UrlRequestPtr GetCatalogUrlRequestBuilder::Build() {
 ///////////////////////////////////////////////////////////////////////////////
 
 std::string GetCatalogUrlRequestBuilder::BuildUrl() const {
+  // TODO(Moritz Haller): Bump catalog version
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_server/get_catalog_url_request_builder.cc:30]: Discussed with @moritzhaller and this change will be made once the server side is ready. There are a few places that need changing.

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ads_database_table.cc [hunk 2]
```diff
@@ -136,6 +143,21 @@ void CreativeAds::MigrateToV10(DBTransaction* transaction) {
   CreateTableV10(transaction);
 }
 
+void CreativeAds::MigrateToV12(DBTransaction* transaction) {
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ads_database_table.cc:146]: v10 code can be removed and we can just rebuild the table in v12 for catalog changees

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  base::FieldTrial* field_trial = base::FieldTrialList::Find(study_name);
+  if (!field_trial) {
+    return base::nullopt;
+  }
+
+  return field_trial->group_name();
+}
+}  // namespace
+
+// TODO(Moritz Haller): rename to SplitTestingFrequencyCap
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc:33]: You have renamed :-)

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#include "base/strings/stringprintf.h"
+#include "base/time/time.h"
+#include "bat/ads/internal/bundle/creative_ad_info.h"
+#include "bat/ads/internal/frequency_capping/frequency_capping_util.h"
+#include "bat/ads/internal/time_util.h"
+
+namespace ads {
+
+namespace {
+const char kStudyName[] = "AdvertiserSplitTestingStudy";
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc:21]: nit: add empty line after `namespace {`

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+base::Optional<std::string> GetSplitTestingGroup(const std::string study_name) {
+  base::FieldTrial* field_trial = base::FieldTrialList::Find(study_name);
+  if (!field_trial) {
+    return base::nullopt;
+  }
+
+  return field_trial->group_name();
+}
+}  // namespace
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc:31]: nit: add empty line above `}  // namespace`

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.h"
+
+#include <iostream>
+
+#include "base/metrics/field_trial.h"
+#include "base/optional.h"
+#include "base/strings/stringprintf.h"
+#include "base/time/time.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc:13]: unused

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+#include "bat/ads/internal/bundle/creative_ad_info.h"
+#include "bat/ads/internal/frequency_capping/frequency_capping_util.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc:15]: unused

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+#include "bat/ads/internal/time_util.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc:16]: unused

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_ADS_SRC_BAT_ADS_INTERNAL_FREQUENCY_CAPPING_EXCLUSION_RULES_SPLIT_TESTING_FREQUENCY_CAP_H_
+#define BRAVE_VENDOR_BAT_NATIVE_ADS_SRC_BAT_ADS_INTERNAL_FREQUENCY_CAPPING_EXCLUSION_RULES_SPLIT_TESTING_FREQUENCY_CAP_H_
+
+#include <string>
+
+#include "bat/ads/internal/bundle/creative_ad_info.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.h:11]: not needed as forward declared below `struct CreativeAdInfo`

**Review by tmancey — APPROVED** [review]
LGTM++

**Comment by tmancey** [comment]
We need to update the catalog schema to match v7

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ads_database_table.cc (hunk 2 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/database/tables/creative_ads_database_table.cc:134]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc (hunk 3 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc]: fixed :)

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc (hunk 4 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc (hunk 5 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc (hunk 6 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc]: deleted

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc (hunk 7 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc]: deleted

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc (hunk 8 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.cc]: deleted

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.h (hunk 9 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_testing_frequency_cap.h]: deleted

**Comment by moritzhaller** [comment]
CI audit failing with known unrelated vulnerability

**Review by tmancey — APPROVED** [review]
LGTM++

**Review by tmancey — APPROVED** [review]
LGTM++

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_test_frequency_cap.cc [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+}  // namespace
+
+SplitTestFrequencyCap::SplitTestFrequencyCap() = default;
+
+SplitTestFrequencyCap::~SplitTestFrequencyCap() = default;
+
+bool SplitTestFrequencyCap::ShouldExclude(const CreativeAdInfo& ad) {
+  if (!DoesRespectCap(ad)) {
+    last_message_ = base::StringPrintf(
```
Comment (diracdeltas) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_test_frequency_cap.cc:36]: does this get logged to disk in any circumstances? we probably don't want to log creative set IDs if they reveal user's browsing interests.

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_test_frequency_cap.cc [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+#include "base/metrics/field_trial.h"
+#include "base/optional.h"
+#include "base/strings/stringprintf.h"
+#include "bat/ads/internal/bundle/creative_ad_info.h"
+
+namespace ads {
+
+namespace {
+
+const char kStudyName[] = "AdvertiserSplitTestStudy";
```
Comment (diracdeltas) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_test_frequency_cap.cc:17]: i suggest starting a wiki page documenting what each of these experiments is for

Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_test_frequency_cap.cc:17]: @diracdeltas I'll create one together with the blog post and will reference from there.

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_test_frequency_cap.cc (hunk 10 above)
Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/split_test_frequency_cap.cc:36]: @diracdeltas this is logged to console/disk only if the user enabled logging for rewards via cli args

**Pull request merged**